src/java.base/share/classes/javax/security/auth/Subject.java

```diff
@@ -319,6 +329,181 @@ public Subject run() {
         });
     }
 
+
+    private static enum SubjectStorage {
+        ACC, THREAD_LOCAL, SCOPE_LOCAL
+    };
+
+    private static final SubjectStorage SUBJECT_STORAGE
+            = SharedSecrets.getJavaLangAccess().systemAllowSecurityManager()
+                    ? SubjectStorage.ACC
+                    : SubjectStorage.THREAD_LOCAL;
+
+//    private static final ScopeLocal<Subject> SUBJECT_SCOPE
+//            = ScopeLocal.inheritableForType(Subject.class);
+
+    private static final InheritableThreadLocal<Subject> SUBJECT_THREAD_LOCAL =
+            new InheritableThreadLocal<Subject>() {
+                @Override protected Subject initialValue() {
+                    return null;
+                }
+            };
+
+    /**
+     * Return the current installed subject.
+     * <p>
+     * The current installed subject is installed by the {@link #call}
+     * or {@link #run} method. When either {@code call(subject, action)} or
+     * {@code run(subject, action)} is called, {@code action} is executed
+     * with {@code subject} as its current installed subject which can be
+     * retrieved by this method. After {@code action} is finished, the current
+     * installed subject is reset to its previous value. The current installed
+     * subject is {@code null} before the first call of {@code call()}
+     * or {@code run()}.
+     * <p>
+     * When a new thread is created, its current installed subject
+     * is the same as the one of its creator, and will not change even if
+     * its creator's current installed subject is changed to another value.
+     *
+     * @implNote
+     * In this implementation, if a {@code SecurityManager} is allowed,
+     * this method returns the same value as
+     * {@code getSubject(AccessController.getContext())}. Otherwise, the current
+     * installed subject is stored in an inheritable {@code ThreadLocal}.
+     * Also, each of the various {@code doAs(subject, action)} and
```

Comment:
It's not clear if the "Also ..." sentence is only for the non-SM case. I think you could just use one sentence here, replace "Also, each" with "and each ...".

Reply:
Oops, I think I made a mistake here. No matter if there's SM, `current` and `getSubject` returns the same thing. I should have said "the current subject" is stored inside "current AccessControlContext" if SM is on or a `ThreadLocal` is not. I'll rephrase.